### PR TITLE
Remove imports of k8s.io/client-go/pkg/api

### PR DESCRIPTION
**Pull request by caesarxuchao**
k8s.io/client-go/pkg/api will be eliminated by #44784, so we need to remove all imports of it.

This PR is large, but mostly simple changes, like:
* using constants defined in the v1 package instead of those in the internal package 
* using k8s.io/client-go/kubernetes/scheme.Codecs/Schemes instead of those in k8s.io/client-go/pkg/api
* using `v1.SchemeGroupVersion` instead of `api.Registry.GroupOrDie(api.GroupName).GroupVersion` in tests
* importing `k8s.io/kubernetes/pkg/api` instead of `k8s.io/client-go/pkg/api` if possible

With this PR https://github.com/kubernetes/kubernetes/pull/45159, the only remaining imports of `k8s.io/client-go/pkg/api` are from k8s.io/metrics.
``` bash
$ git grep "\"k8s.io/client-go/pkg/api\"" | grep -v "staging/src/k8s.io/client-go/pkg/api"
staging/src/k8s.io/metrics/pkg/apis/custom_metrics/install/install.go:25:	"k8s.io/client-go/pkg/api"
Binary file staging/src/k8s.io/metrics/pkg/apis/custom_metrics/types.generated.go matches
staging/src/k8s.io/metrics/pkg/apis/custom_metrics/types.go:22:	"k8s.io/client-go/pkg/api"
staging/src/k8s.io/metrics/pkg/apis/metrics/install/install.go:26:	"k8s.io/client-go/pkg/api"
staging/src/k8s.io/metrics/pkg/apis/metrics/types.go:21:	"k8s.io/client-go/pkg/api"
Binary file staging/src/k8s.io/metrics/pkg/apis/metrics/v1alpha1/zz_generated.conversion.go matches
Binary file staging/src/k8s.io/metrics/pkg/apis/metrics/zz_generated.deepcopy.go matches
staging/src/k8s.io/metrics/pkg/client/clientset_generated/clientset/fake/clientset_generated.go:24:	"k8s.io/client-go/pkg/api"
staging/src/k8s.io/metrics/pkg/client/custom_metrics/client.go:26:	"k8s.io/client-go/pkg/api"
staging/src/k8s.io/metrics/pkg/client/custom_metrics/fake/fake_client.go:24:	"k8s.io/client-go/pkg/api"
```

### CODE REVIEW

**Comment by caesarxuchao** [comment]
@k8s-bot cvm gce e2e test this

**Comment by caesarxuchao** [comment]
Tests passed. I'll try remove the import_known_versions.go from clientset, and see if any tests will fail.

File: pkg/api/v1/types.go [hunk 1]
```diff
@@ -4428,3 +4428,44 @@ type NodeResources struct {
...
 }
+
+// These constants are for remote command execution and port forwarding and are
+// used by both the client side and server side components.
+//
+// This is probably not the ideal place for them, but it didn't seem worth it
+// to create pkg/exec and pkg/portforward just to contain a single file with
+// constants in it.  Suggestions for more appropriate alternatives are
+// definitely welcome!
+const (
```
Comment (caesarxuchao) [inline pkg/api/v1/types.go]: These are copied from pkg/api/types.go.

File: staging/src/k8s.io/client-go/rest/client_test.go [hunk 2]
```diff
@@ -50,12 +49,13 @@ type TestParam struct {
 	testBodyErrorIsNotNil bool
 }
 
-// TestSerializer makes sure that you're always able to decode an unversioned API object
+// TestSerializer makes sure that you're always able to decode metav1.Status
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/rest/client_test.go]: Some of the changes in this file seem to follow the pattern of the rest of the changes but there's some that don't, why?

File: staging/src/k8s.io/client-go/rest/config_test.go [hunk 3]
```diff
@@ -139,13 +139,14 @@ func TestDefaultKubernetesUserAgent(t *testing.T) {
 }
 
 func TestRESTClientRequires(t *testing.T) {
-	if _, err := RESTClientFor(&Config{Host: "127.0.0.1", ContentConfig: ContentConfig{NegotiatedSerializer: api.Codecs}}); err == nil {
+	gvCopy := v1.SchemeGroupVersion
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/rest/config_test.go]: if this is just about making the following lines shorter, I suggest calling this `sgv`. If it's actually about making a copy because it might get mutated, I'd make a function so that each RESTClientFor line gets its own copy.

File: staging/src/k8s.io/client-go/rest/request_test.go [hunk 4]
```diff
@@ -94,9 +93,10 @@ func TestRequestSetsHeaders(t *testing.T) {
 }
 
 func TestRequestWithErrorWontChange(t *testing.T) {
+	gvCopy := v1.SchemeGroupVersion
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/rest/request_test.go]: Yeah I don't get why there is this copy line.

File: staging/src/k8s.io/client-go/rest/watch/decoder_test.go [hunk 5]
```diff
@@ -25,28 +25,31 @@ import (
...
 )
 
 func TestDecoder(t *testing.T) {
 	table := []watch.EventType{watch.Added, watch.Deleted, watch.Modified, watch.Error}
 
 	for _, eventType := range table {
 		out, in := io.Pipe()
-		codec := api.Codecs.LegacyCodec(v1.SchemeGroupVersion)
+
+		jsonSerializer := runtimejson.NewSerializer(runtimejson.DefaultMetaFactory, scheme.Scheme, scheme.Scheme, false)
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/rest/watch/decoder_test.go]: I don't understand why you can't call `scheme.Codecs.LegacyCodec`?

File: staging/src/k8s.io/client-go/rest/watch/encoder_test.go [hunk 6]
```diff
@@ -24,49 +24,48 @@ import (
...
-			api.Codecs.LegacyCodec(v1.SchemeGroupVersion),
+			&v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "foo"}},
 		},
 	}
 	for i, testCase := range testCases {
 		buf := &bytes.Buffer{}
 
-		codec := testCase.Codec
-		encoder := restclientwatch.NewEncoder(streaming.NewEncoder(buf, codec), codec)
+		jsonSerializer := runtimejson.NewSerializer(runtimejson.DefaultMetaFactory, scheme.Scheme, scheme.Scheme, false)
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/rest/watch/encoder_test.go]: It would be better to make a function than copy-pasting these three lines in multiple places :)

File: staging/src/k8s.io/client-go/tools/record/event_test.go [hunk 7]
```diff
@@ -631,15 +641,18 @@ func TestMultiSinkCache(t *testing.T) {
 		},
 	}
 	testPod2 := &v1.Pod{
+		TypeMeta: metav1.TypeMeta{
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/tools/record/event_test.go]: Why add this? If it's because you want to pass nil for a scheme, maybe think about making/finding a scheme instead... I'll be more comfortable with minimal changes to test code in a change like this--not sure if this change to the test leaves it sufficiently similar to actual uses to remain useful...

**Review by lavalamp — CHANGES_REQUESTED** [review]
I would title this PR something more like "Make references of api.Scheme refer to the appropriate scheme"

File: staging/src/k8s.io/apimachinery/pkg/api/meta/help.go [hunk 8]
```diff
@@ -175,6 +175,9 @@ func SetList(list runtime.Object, objects []runtime.Object) error {
 	slice := reflect.MakeSlice(items.Type(), len(objects), len(objects))
 	for i := range objects {
 		dest := slice.Index(i)
+		if dest.Type() == reflect.TypeOf(runtime.RawExtension{}) {
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apimachinery/pkg/api/meta/help.go]: A test was using [api.List](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L3372) and this PR changes it to [v1.List](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L3900). `api.List.Items` is slice of `runtime.Object`, while `v1.List.Items` is a slice of `runtime.RawExtension`. This change makes SetList to work for `v1.List`, otherwise it's going to try convert an object to runtime.RawExtension and there is no conversion function to do that.

File: pkg/client/clientset_generated/clientset/typed/core/v1/pod_expansion.go [hunk 9]
```diff
@@ -41,5 +41,5 @@ func (c *pods) Evict(eviction *policy.Eviction) error {
 
 // Get constructs a request for getting the logs for a pod
 func (c *pods) GetLogs(name string, opts *v1.PodLogOptions) *restclient.Request {
-	return c.client.Get().Namespace(c.ns).Name(name).Resource("pods").SubResource("log").VersionedParams(opts, api.ParameterCodec)
+	return c.client.Get().Namespace(c.ns).Name(name).Resource("pods").SubResource("log").VersionedParams(opts, scheme.ParameterCodec)
```
Comment (caesarxuchao) [inline pkg/client/clientset_generated/clientset/typed/core/v1/pod_expansion.go]: Yeah, probably should build the scheme into the restclient.

> You can do that for expediency but it still needs to be a variable / method on the client and this function should get it from there.

I don't want to do that at the moment because client-gen is using scheme.ParameterCodec everywhere: https://github.com/kubernetes/kubernetes/blob/master/pkg/client/clientset_generated/clientset/typed/core/v1/pod.go#L119

File: staging/src/k8s.io/client-go/rest/client_test.go [hunk 10]
```diff
@@ -50,12 +49,13 @@ type TestParam struct {
...
 func TestSerializer(t *testing.T) {
+	gv := v1beta1.SchemeGroupVersion
 	contentConfig := ContentConfig{
 		ContentType:          "application/json",
-		GroupVersion:         &schema.GroupVersion{Group: "other", Version: runtime.APIVersionInternal},
-		NegotiatedSerializer: api.Codecs,
+		GroupVersion:         &gv,
+		NegotiatedSerializer: serializer.DirectCodecFactory{CodecFactory: scheme.Codecs},
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/rest/client_test.go]: This is the actual codec that's used by the [clientset](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/kubernetes/typed/core/v1/core_client.go#L147). The original code is wrong.

File: staging/src/k8s.io/client-go/rest/client_test.go [hunk 11]
```diff
@@ -93,7 +93,7 @@ func TestDoRequestFailed(t *testing.T) {
 		Message: " \"\" not found",
 		Details: &metav1.StatusDetails{},
 	}
-	expectedBody, _ := runtime.Encode(api.Codecs.LegacyCodec(v1.SchemeGroupVersion), status)
+	expectedBody, _ := runtime.Encode(scheme.Codecs.LegacyCodec(v1.SchemeGroupVersion), status)
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/rest/client_test.go]: The codec we get from `codec := c.LegacyCodec(version)` is going to encode to `version` and decode to `internalVersion`. So using it to encode is fine, but to decode, we need to extract the serializer part or wrap the scheme.Codec into a DirectCodec, which is used by the clientset.

File: staging/src/k8s.io/client-go/rest/config_test.go (hunk 3 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/rest/config_test.go]: It's to prevent it from getting mutated. Added the function as suggested :)

File: staging/src/k8s.io/client-go/rest/watch/encoder_test.go (hunk 6 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/rest/watch/encoder_test.go]: Done.

File: staging/src/k8s.io/client-go/tools/record/event_test.go (hunk 7 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/tools/record/event_test.go]: You are right. The test should assume TypeMeta is empty in memory. Now i reverted the addition and changed to use scheme.Scheme.

File: staging/src/k8s.io/client-go/rest/watch/encoder_test.go [hunk 12]
```diff
@@ -24,49 +24,51 @@ import (
...
 	"k8s.io/apimachinery/pkg/watch"
-	"k8s.io/client-go/pkg/api"
+	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/pkg/api/v1"
 	restclientwatch "k8s.io/client-go/rest/watch"
-
-	_ "k8s.io/client-go/pkg/api/install"
 )
 
+// getEncoder mimics how k8s.io/client-go/rest.createSerializers creates a encoder
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/rest/watch/encoder_test.go]: Is there a reason not to call that directly?

File: staging/src/k8s.io/client-go/rest/client_test.go [hunk 13]
```diff
@@ -231,12 +231,13 @@ func validate(testParam TestParam, t *testing.T, body []byte, fakeHandler *utilt
...
+	statusOut, err := runtime.Decode(scheme.Codecs.UniversalDeserializer(), body)
 	if testParam.testBody {
-		if testParam.testBodyErrorIsNotNil {
-			if err == nil {
-				t.Errorf("Expected Error")
-			}
+		if testParam.testBodyErrorIsNotNil && err == nil {
+			t.Errorf("Expected Error")
+		}
+		if !testParam.testBodyErrorIsNotNil && err != nil {
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/rest/client_test.go]: I take it the test logic was just wrong previously?

File: staging/src/k8s.io/client-go/rest/client_test.go [hunk 14]
```diff
@@ -329,11 +330,12 @@ func testServerEnv(t *testing.T, statusCode int) (*httptest.Server, *utiltesting
 }
 
 func restClient(testServer *httptest.Server) (*RESTClient, error) {
+	gvCopy := v1.SchemeGroupVersion
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/rest/client_test.go]: s/gvCopy/sgv/g

File: staging/src/k8s.io/client-go/rest/config_test.go [hunk 15]
```diff
@@ -139,13 +139,17 @@ func TestDefaultKubernetesUserAgent(t *testing.T) {
 }
 
 func TestRESTClientRequires(t *testing.T) {
-	if _, err := RESTClientFor(&Config{Host: "127.0.0.1", ContentConfig: ContentConfig{NegotiatedSerializer: api.Codecs}}); err == nil {
+	getGV := func() *schema.GroupVersion {
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/rest/config_test.go]: add comment: "this is here since RESTClient might mutate the value we pass in"

(Is this true? It seems weird, why does RESTClient do this?)

File: staging/src/k8s.io/client-go/rest/request_test.go [hunk 16]
```diff
@@ -94,9 +93,10 @@ func TestRequestSetsHeaders(t *testing.T) {
...
 	original := Request{
 		err:     errors.New("test"),
-		content: ContentConfig{GroupVersion: &api.Registry.GroupOrDie(api.GroupName).GroupVersion},
+		content: ContentConfig{GroupVersion: &gvCopy},
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/rest/request_test.go]: It's much better to get these things from a registry than to get them directly from the package. Can it be gotten from a registry living in apimachinery somewhere?

**Review by lavalamp — COMMENTED** [review]
A few more thoughts

File: staging/src/k8s.io/client-go/rest/watch/encoder_test.go (hunk 12 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/rest/watch/encoder_test.go]: it's in a different package and this test is a black-box one.

File: staging/src/k8s.io/client-go/rest/client_test.go (hunk 13 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/rest/client_test.go]: Right, it was wrong and made the test hard to debug.

File: staging/src/k8s.io/client-go/rest/config_test.go (hunk 15 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/rest/config_test.go]: It's not mutating. I'm just trying to "prevent" potential mutation. Shall i just revert the change?

File: staging/src/k8s.io/client-go/rest/request_test.go (hunk 16 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/rest/request_test.go]: There is no existing Registry in apimachinery.

IMO client should have a global test registry shared by all client tests. And if a client test needs aspecial registry, it creates its own.

**Comment by caesarxuchao** [comment]
@k8s-bot node e2e test this

**Comment by caesarxuchao** [comment]
@lavalamp could you take another look?

I added a commit that removed more dependency on client-go/pkg/api since you reviewed last time. We are now very close to remove all such dependencies.

File: staging/src/k8s.io/client-go/rest/config_test.go (hunk 15 above)
Comment (lavalamp) [inline staging/src/k8s.io/client-go/rest/config_test.go]: Yeah I would try to keep this change focused.

File: staging/src/k8s.io/client-go/rest/request_test.go (hunk 16 above)
Comment (lavalamp) [inline staging/src/k8s.io/client-go/rest/request_test.go]: I do agree tests should probably be constructing their own registry.

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go [hunk 17]
```diff
@@ -109,11 +109,24 @@ func (w *WebhookTokenAuthenticator) AuthenticateToken(token string) (user.Info,
 	}, true, nil
 }
 
+// NOTE: client-go doesn't provide a registry. client-go does registers the
+// authentication/v1beta1. We construct a registry that acknowledges
+// authentication/v1beta1 as an enabled version to pass the test in
+// NetGenericWebhook.
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go]: I am confused about the mention of the test when this appears to be referenced in real code?

File: staging/src/k8s.io/client-go/rest/client_test.go (hunk 14 above)
Comment (lavalamp) [inline staging/src/k8s.io/client-go/rest/client_test.go]: This comment still applies

**Comment by lavalamp** [comment]
just some minor comments left.

File: staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go [hunk 18]
```diff
@@ -207,11 +209,24 @@ func convertToSARExtra(extra map[string][]string) map[string]authorization.Extra
 	return ret
 }
 
+// NOTE: client-go doesn't provide a registry. client-go does registers the
+// authorization/v1beta1. We construct a registry that acknowledges
+// authorization/v1beta1 as an enabled version to pass a check enforced in
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go]: Updated the comment to be more accurate.

**Comment by caesarxuchao** [comment]
Comments addressed. PTAL @lavalamp. Thanks!

**Review by lavalamp — APPROVED** [review]
